runtime/compiler/optimizer/JProfilingRecompLoopTest.hpp

```diff
@@ -0,0 +1,57 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#include <stdint.h>                           // for int32_t
+#include "optimizer/Optimization.hpp"         // for Optimization
+#include "optimizer/OptimizationManager.hpp"  // for OptimizationManager
+#include "infra/Checklist.hpp"                // fir NodeChecklist
+#include "runtime/J9Profiler.hpp"
+#include "optimizer/Structure.hpp"
+/**
+ * Class TR_JProfilingRecompLoopTest
+ * ===================================
+ *
+ * When jProfiling in compilations are enabled, we have only one recompilation
+ * test that counts the method invocation to trip up recompilation. For work loads
+ * where we are spending a lot of time in loops, we are hitting a risk where we
+ * are running into compiled method body with profiling trees if we just rely on
+ * method invocations. This optimization scans the method body and puts 
+ * recompilation test after asynccheck node to consider number of times a loop
+ * is running.
+ */
+class TR_JProfilingRecompLoopTest : public TR::Optimization
+   {
+   public:
+   typedef TR::typed_allocator<std::pair<std::pair<TR::TreeTop*, TR::Block*>, int32_t>, TR::Region &> RecompilationTestLocationAllocator;
```

Comment:
can you typedef the std::pair<TR::TreeTop*, TR::Block*> to some kind of helpful name and use it in the typedefs here to make this code easier to understand - what does this pair represent conceptually?

Reply:
I am collecting test location information here that incorporates TreeTop where we want to add test (asynccheck) , enclosing block (instead of calling enclosing block we get that information and store it while walking treetops ) and loop nesting depth. 
The reason I need this information here is because before we add the recompilation test and split the blocks, I set the structure to NULL and I lost information about loop nesting. 
Commenting for, https://github.com/eclipse/openj9/pull/2348#pullrequestreview-135848068
Once this is merged, I will do analysis on what is more costly -> Setting structure to NULL before splitting or fixing up structure every time we add test.
If later is more beneficial, I can clean up depth information from testLocationInfo.